File: 0450.Delete Node in a BST/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(root == nullptr){
            return nullptr;
        }
        if(key < root->val){
            root->left = deleteNode(root->left, key);
        }else if(key > root->val){
            root->right = deleteNode(root->right, key);
        }else{
            if(root->left == nullptr)
                return root->right;
            if(root->right == nullptr)
                return root->left;
            TreeNode* min = findMin(root->right);
            root->val = min->val;
            root->right = deleteNode(root->right, root->val);
        }
        return root;
    }
    
    
    TreeNode* findMin(TreeNode* root){
        while(root->left){
            root = root->left;
        }
        return root;
    }
};

File: 0450.Delete Node in a BST/tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def deleteNode(root: TreeNode, key: Int): TreeNode = {
        if(root == null)
            return null;
        if(key < root.value){
            root.left = deleteNode(root.left, key);
        }else if(key > root.value){
            root.right = deleteNode(root.right, key)
        }else{
            if(root.left == null)
                return root.right;
            if(root.right == null)
                return root.left;
            var minNode: TreeNode = findMin(root.right, key);
            root.value = minNode.value;
            root.right = deleteNode(root.right, root.value);
        }
        return root;
    }
    
    def findMin(root: TreeNode, key: Int): TreeNode = {
        var node = root;
        while(node.left != null){
            node = node.left;
        }
        return node;
    }
}